File: platform_src/ios/Elastos/Classes/ManagerDBAdapter.swift
 /*
  * Copyright (c) 2018 Elastos Foundation
  *
  * Permission is hereby granted, free of charge, to any person obtaining a copy
  * of this software and associated documentation files (the "Software"), to deal
  * in the Software without restriction, including without limitation the rights
  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  * copies of the Software, and to permit persons to whom the Software is
  * furnished to do so, subject to the following conditions:
  *
  * The above copyright notice and this permission notice shall be included in all
  * copies or substantial portions of the Software.
  *
  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */

import Foundation
import SQLite

 class ManagerDBAdapter {

    @objc static let DATABASE_NAME = "manager.db";
    @objc static let VERSION = 1;
    @objc static let AUTH_PLUGIN_TABLE = "auth_plugin";
    @objc static let AUTH_URL_TABLE = "auth_url";
    @objc static let AUTH_INTENT_TABLE = "auth_intent";
    @objc static let ICONS_TABLE = "icons";
    @objc static let LOCALE_TABLE = "locale";
    @objc static let FRAMEWORK_TABLE = "framework";
    @objc static let PLATFORM_TABLE = "platform";
    @objc static let INTENT_FILTER_TABLE = "intent_filter";
    @objc static let APP_TABLE = "app";

    let db: Connection;
    let tid = Expression<Int64>(AppInfo.TID)
    let app_tid = Expression<Int64>(AppInfo.APP_TID);

    let app_id = Expression<String>(AppInfo.APP_ID)
    let version = Expression<String>(AppInfo.VERSION)
    let version_code = Expression<Int>(AppInfo.VERSION_CODE)
    let name = Expression<String>(AppInfo.NAME)
    let short_name = Expression<String?>(AppInfo.SHORT_NAME)
    let description = Expression<String?>(AppInfo.DESCRIPTION)
    let start_url = Expression<String>(AppInfo.START_URL)
    let start_visible = Expression<String>(AppInfo.START_VISIBLE)
    let author_name = Expression<String?>(AppInfo.AUTHOR_NAME)
    let author_email = Expression<String?>(AppInfo.AUTHOR_EMAIL)
    let default_locale = Expression<String?>(AppInfo.DEFAULT_LOCAL)
    let background_color = Expression<String?>(AppInfo.BACKGROUND_COLOR)
    let theme_display = Expression<String?>(AppInfo.THEME_DISPLAY)
    let theme_color = Expression<String?>(AppInfo.THEME_COLOR)
    let theme_font_name = Expression<String?>(AppInfo.THEME_FONT_NAME)
    let theme_font_color = Expression<String?>(AppInfo.THEME_FONT_COLOR)
    let install_time = Expression<Int64>(AppInfo.INSTALL_TIME)
    let built_in = Expression<Bool>(AppInfo.BUILT_IN)
    let remote = Expression<Bool>(AppInfo.REMOTE)
    let launcher = Expression<Bool>(AppInfo.LAUNCHER)
    let category = Expression<String?>(AppInfo.CATEGORY)
    let key_words = Expression<String?>(AppInfo.KEY_WORDS)

    let plugin = Expression<String>(AppInfo.PLUGIN)
    let url = Expression<String>(AppInfo.URL)
    let authority = Expression<Int>(AppInfo.AUTHORITY)

    let src = Expression<String>(AppInfo.SRC)
    let sizes = Expression<String>(AppInfo.SIZES)
    let type = Expression<String>(AppInfo.TYPE)

    let language = Expression<String>(AppInfo.LANGUAGE)

    let action = Expression<String>(AppInfo.ACTION)

    let plugins = Table(ManagerDBAdapter.AUTH_PLUGIN_TABLE)
    let urls = Table(ManagerDBAdapter.AUTH_URL_TABLE)
    let intents = Table(ManagerDBAdapter.AUTH_INTENT_TABLE)
    let icons = Table(ManagerDBAdapter.ICONS_TABLE)
    let locales = Table(ManagerDBAdapter.LOCALE_TABLE)
    let frameworks = Table(ManagerDBAdapter.FRAMEWORK_TABLE)
    let platforms = Table(ManagerDBAdapter.PLATFORM_TABLE)
    let intent_filters = Table(ManagerDBAdapter.INTENT_FILTER_TABLE)
    let apps = Table(ManagerDBAdapter.APP_TABLE)

    init(_ dataPath: String) {
        let path = dataPath + ManagerDBAdapter.DATABASE_NAME;
        db = try! Connection(path)
        try! creatTables();
    }

    func creatTables() throws {
        try db.run(plugins.create(ifNotExists: true) { t in
            t.column(tid, primaryKey: .autoincrement)
            t.column(app_tid)
            t.column(plugin)
            t.column(authority)
        })

        try db.run(urls.create(ifNotExists: true) { t in
            t.column(tid, primaryKey: .autoincrement)
            t.column(app_tid)
            t.column(url)
            t.column(authority)
        })
        
        try db.run(intents.create(ifNotExists: true) { t in
            t.column(tid, primaryKey: .autoincrement)
            t.column(app_tid)
            t.column(url)
            t.column(authority)
        })

        try db.run(icons.create(ifNotExists: true) { t in
            t.column(tid, primaryKey: .autoincrement)
            t.column(app_tid)
            t.column(src)
            t.column(sizes)
            t.column(type)
        })

        try db.run(locales.create(ifNotExists: true) { t in
            t.column(tid, primaryKey: .autoincrement)
            t.column(app_tid)
            t.column(language)
            t.column(name)
            t.column(short_name)
            t.column(description)
            t.column(author_name)
        })

        try db.run(frameworks.create(ifNotExists: true) { t in
            t.column(tid, primaryKey: .autoincrement)
            t.column(app_tid)
            t.column(name)
            t.column(version)
        })

        try db.run(platforms.create(ifNotExists: true) { t in
            t.column(tid, primaryKey: .autoincrement)
            t.column(app_tid)
            t.column(name)
            t.column(version)
        })

        try db.run(intent_filters.create(ifNotExists: true) { t in
            t.column(tid, primaryKey: .autoincrement)
            t.column(app_id)
            t.column(action)
        })

        try db.run(apps.create(ifNotExists: true) { t in
            t.column(tid, primaryKey: .autoincrement)
            t.column(app_id, unique: true)
            t.column(version)
            t.column(version_code)
            t.column(name)
            t.column(short_name)
            t.column(description)
            t.column(start_url)
            t.column(start_visible)
            t.column(type)
            t.column(author_name)
            t.column(author_email)
            t.column(default_locale)
            t.column(background_color)
            t.column(theme_display)
            t.column(theme_color)
            t.column(theme_font_name)
            t.column(theme_font_color)
            t.column(install_time)
            t.column(built_in)
            t.column(remote)
            t.column(launcher)
            t.column(category)
            t.column(key_words)
        })
    }

    func dropTables() throws {
        try db.run(plugins.drop());
        try db.run(urls.drop());
        try db.run(icons.drop());
        try db.run(locales.drop());
        try db.run(frameworks.drop());
        try db.run(platforms.drop());
        try db.run(intent_filters.drop());
        try db.run(apps.drop());
    }

    func addAppInfo(_ info: AppInfo) throws {
        try db.transaction {
            info.tid = try db.run(apps.insert(
                app_id <- info.app_id,
                version <- info.version,
                version_code <- info.version_code,
                name <- info.name,
                short_name <- info.short_name,
                description <- info.desc,
                start_url <- info.start_url,
                start_visible <- info.start_visible,
                type <- info.type,
                author_name <- info.author_name,
                author_email <- info.author_email,
                default_locale <- info.default_locale,
                background_color <- info.background_color,
                theme_display <- info.theme_display,
                theme_color <- info.theme_color,
                theme_font_name <- info.theme_font_name,
                theme_font_color <- info.theme_font_color,
                install_time <- info.install_time,
                built_in <- info.built_in,
                remote <- info.remote,
                launcher <- info.launcher,
                category <- info.category,
                key_words <- info.key_words
            ));

            print("inserted id: \(info.tid)")

            for icon in info.icons {
                try db.run(icons.insert(app_tid <- info.tid,
                        src <- icon.src,
                        sizes <- icon.sizes,
                        type <- icon.type));
            }

            for pluginAuth in info.plugins {
                try db.run(plugins.insert(app_tid <- info.tid,
                        plugin <- pluginAuth.plugin,
                        authority <- pluginAuth.authority));
            }

            for urlAuth in info.urls {
                try db.run(urls.insert(app_tid <- info.tid,
                        url <- urlAuth.url,
                        authority <- urlAuth.authority));
            }

            for urlAuth in info.intents {
                try db.run(intents.insert(app_tid <- info.tid,
                        url <- urlAuth.url,
                        authority <- urlAuth.authority));
            }
            
            for locale in info.locales {
                try db.run(locales.insert(app_tid <- info.tid,
                                       language <- locale.language,
                                       name <- locale.name,
                                       short_name <- locale.short_name,
                                       description <- locale.desc,
                                       author_name <- locale.author_name));
            }

            for framework in info.frameworks {
                try db.run(frameworks.insert(app_tid <- info.tid,
                                        name <- framework.name,
                                        version <- framework.version));
            }

            for platform in info.platforms {
                try db.run(platforms.insert(app_tid <- info.tid,
                                        name <- platform.name,
                                        version <- platform.version));
            }
        }
    }

    func getInfos(_ rows: AnySequence<Row>) throws -> [AppInfo] {
        var infos = [AppInfo]();
        for app in rows {
            let info = AppInfo();
            info.tid = app[tid];
            info.app_id = app[app_id];
            info.version = app[version];
            info.version_code = app[version_code];
            info.name = app[name];
            info.short_name = app[short_name] ?? "";
            info.desc = app[description] ?? "";
            info.start_url = app[start_url];
            info.start_visible = app[start_visible] ?? "";
            info.type = app[type];
            info.author_name = app[author_name] ?? "";
            info.author_email = app[author_email] ?? "";
            info.default_locale = app[default_locale] ?? "";
            info.background_color = app[background_color] ?? "";
            info.theme_display = app[theme_display] ?? "";
            info.theme_font_name = app[theme_font_name] ?? "";
            info.theme_font_color = app[theme_font_color] ?? "";
            info.install_time = app[install_time];
            info.built_in = app[built_in];
            info.remote = app[remote];
            info.launcher = app[launcher];
            info.category = app[category] ?? "other";
            info.key_words = app[key_words] ?? "";

            for icon in try db.prepare(icons.select(*).filter(app_tid == info.tid)) {
                info.addIcon(icon[src], icon[sizes], icon[type]);
            }

            for pluginAuth in try db.prepare(plugins.select(*).filter(app_tid == info.tid)) {
                let name = pluginAuth[plugin];
                let auth = pluginAuth[authority]
                info.addPlugin(name, auth);
            }

            for urlAuth in try db.prepare(urls.select(*).filter(app_tid == info.tid)) {
                info.addUrl(urlAuth[url], urlAuth[authority]);
            }

            for urlAuth in try db.prepare(intents.select(*).filter(app_tid == info.tid)) {
                info.addIntent(urlAuth[url], urlAuth[authority]);
            }
            
            for locale in try db.prepare(locales.select(*).filter(app_tid == info.tid)) {
                info.addLocale(locale[language], locale[name], app[short_name] ?? "", locale[description] ?? "", locale[author_name] ?? "");
            }

            for framework in try db.prepare(frameworks.select(*).filter(app_tid == info.tid)) {
                info.addFramework(framework[name], framework[version]);
            }

            for platform in try db.prepare(platforms.select(*).filter(app_tid == info.tid)) {
                info.addPlatform(platform[name], platform[version]);
            }

            infos.append(info);
        }
        return infos;
    }

    func getAppInfo(_ id: String) throws -> AppInfo? {
        let query = apps.select(*)
            .filter(app_id == id && !launcher)
        let rows = try db.prepare(query);
        let infos = try getInfos(rows);
        guard infos.count > 0 else {
            return nil;
        }
        return infos[0];
    }

    func getAppInfos() throws -> [AppInfo] {
        let query = apps.select(*)
            .filter(!launcher)
        let rows = try db.prepare(query);
        return try getInfos(rows);
    }

    func getLauncherInfo() throws -> AppInfo? {
        let query = apps.select(*)
            .filter(launcher)
        let rows = try db.prepare(query);
        let infos = try getInfos(rows);
        guard infos.count > 0 else {
            return nil;
        }
        return infos[0];
    }

    func updatePluginAuth(_ item: PluginAuth, _ auth: Int) throws {
        try db.transaction {
            let pluginAuth = plugins.filter(plugin == item.plugin);
            try db.run(pluginAuth.update(authority <- auth));
        }
    }

    func updateUrlAuth(_ item: UrlAuth, _ auth: Int) throws {
        try db.transaction {
            let urlAuth = urls.filter(url == item.url);
            try db.run(urlAuth.update(authority <- auth));
        }
    }

    func removeAppInfo(_ info: AppInfo) throws {
        var items = plugins.filter(app_tid == info.tid);
        try db.run(items.delete());
        items = urls.filter(app_tid == info.tid);
        try db.run(items.delete());
        items = icons.filter(app_tid == info.tid);
        try db.run(items.delete());
        items = locales.filter(app_tid == info.tid);
        try db.run(items.delete());
        items = frameworks.filter(app_tid == info.tid);
        try db.run(items.delete());
        items = platforms.filter(app_tid == info.tid);
        try db.run(items.delete());
        items = intent_filters.filter(app_id == info.app_id);
        try db.run(items.delete());
        items = apps.filter(tid == info.tid);
        try db.run(items.delete());
    }

    func addIntentFilter(_ intent: Intent) throws {
        try db.transaction {
            try db.run(intent_filters.insert(app_id <- intent.app_id,
                                      action <- intent.action));
        }
    }

    func getIntentFilter(_ act: String) throws -> [String] {
        let query = intent_filters.select(app_id)
            .filter(action == act)
        let rows = try db.prepare(query);
        var ids = [String]();
        for intent in rows {
            ids.append(intent[app_id]);
        }
        return ids;
    }

    func clean() throws {
        try dropTables();
        try creatTables();
    }

 }

File: platform_src/android/app/src/main/java/org/elastos/trinity/runtime/ManagerDBAdapter.java
 /*
  * Copyright (c) 2018 Elastos Foundation
  *
  * Permission is hereby granted, free of charge, to any person obtaining a copy
  * of this software and associated documentation files (the "Software"), to deal
  * in the Software without restriction, including without limitation the rights
  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  * copies of the Software, and to permit persons to whom the Software is
  * furnished to do so, subject to the following conditions:
  *
  * The above copyright notice and this permission notice shall be included in all
  * copies or substantial portions of the Software.
  *
  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */

package org.elastos.trinity.runtime;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

public class ManagerDBAdapter {
    ManagerDBHelper helper;
    Context context;
    public ManagerDBAdapter(Context context)
    {
        helper = new ManagerDBHelper(context);
        SQLiteDatabase db = helper.getWritableDatabase();
        this.context = context;
    }

    public void clean() {
        SQLiteDatabase db = helper.getWritableDatabase();
        helper.onUpgrade(db, 0, 1);
    }

    public boolean addAppInfo(AppInfo info) {
        if (info != null) {
            SQLiteDatabase db = helper.getWritableDatabase();
            ContentValues contentValues = new ContentValues();
            contentValues.put(AppInfo.APP_ID, info.app_id);
            contentValues.put(AppInfo.VERSION, info.version);
            contentValues.put(AppInfo.NAME, info.name);
            contentValues.put(AppInfo.SHORT_NAME, info.short_name);
            contentValues.put(AppInfo.DESCRIPTION, info.description);
            contentValues.put(AppInfo.START_URL, info.start_url);
            contentValues.put(AppInfo.TYPE, info.type);
            contentValues.put(AppInfo.AUTHOR_NAME, info.author_name);
            contentValues.put(AppInfo.AUTHOR_EMAIL, info.author_email);
            contentValues.put(AppInfo.DEFAULT_LOCAL, info.default_locale);
            contentValues.put(AppInfo.BACKGROUND_COLOR, info.background_color);
            contentValues.put(AppInfo.THEME_DISPLAY, info.theme_display);
            contentValues.put(AppInfo.THEME_COLOR, info.theme_color);
            contentValues.put(AppInfo.THEME_FONT_NAME, info.theme_font_name);
            contentValues.put(AppInfo.THEME_FONT_COLOR, info.theme_font_color);
            contentValues.put(AppInfo.INSTALL_TIME, info.install_time);
            contentValues.put(AppInfo.BUILT_IN, info.built_in);
            contentValues.put(AppInfo.REMOTE, info.remote);
            contentValues.put(AppInfo.LAUNCHER, info.launcher);
            long tid = db.insert(ManagerDBHelper.APP_TABLE, null, contentValues);

            if (tid == -1) {
                return false;
            }
            info.tid = tid;

            for (AppInfo.Icon icon : info.icons) {
                contentValues = new ContentValues();
                contentValues.put(AppInfo.APP_TID, tid);
                contentValues.put(AppInfo.SRC, icon.src);
                contentValues.put(AppInfo.SIZES, icon.sizes);
                contentValues.put(AppInfo.TYPE, icon.type);
                db.insert(ManagerDBHelper.ICONS_TABLE, null, contentValues);
            }

            for (AppInfo.PluginAuth pluginAuth : info.plugins) {
                contentValues = new ContentValues();
                contentValues.put(AppInfo.APP_TID, tid);
                contentValues.put(AppInfo.PLUGIN, pluginAuth.plugin);
                contentValues.put(AppInfo.AUTHORITY, pluginAuth.authority);
                db.insert(ManagerDBHelper.AUTH_PLUGIN_TABLE, null, contentValues);
            }

            for (AppInfo.UrlAuth urlAuth : info.urls) {
                contentValues = new ContentValues();
                contentValues.put(AppInfo.APP_TID, tid);
                contentValues.put(AppInfo.URL, urlAuth.url);
                contentValues.put(AppInfo.AUTHORITY, urlAuth.authority);
                db.insert(ManagerDBHelper.AUTH_URL_TABLE, null, contentValues);
            }

            for (AppInfo.Locale locale : info.locales) {
                contentValues = new ContentValues();
                contentValues.put(AppInfo.APP_TID, tid);
                contentValues.put(AppInfo.LANGUAGE, locale.language);
                contentValues.put(AppInfo.NAME, locale.name);
                contentValues.put(AppInfo.SHORT_NAME, locale.short_name);
                contentValues.put(AppInfo.DESCRIPTION, locale.description);
                contentValues.put(AppInfo.AUTHOR_NAME, locale.author_name);
                db.insert(ManagerDBHelper.LACALE_TABLE, null, contentValues);
            }

            for (AppInfo.Framework framework : info.frameworks) {
                contentValues = new ContentValues();
                contentValues.put(AppInfo.APP_TID, tid);
                contentValues.put(AppInfo.NAME, framework.name);
                contentValues.put(AppInfo.VERSION, framework.version);
                db.insert(ManagerDBHelper.FRAMEWORK_TABLE, null, contentValues);
            }


            for (AppInfo.Platform platform : info.platforms) {
                contentValues = new ContentValues();
                contentValues.put(AppInfo.APP_TID, tid);
                contentValues.put(AppInfo.NAME, platform.name);
                contentValues.put(AppInfo.VERSION, platform.version);
                db.insert(ManagerDBHelper.PLATFORM_TABLE, null, contentValues);
            }

            return true;
        }
        else {
            return false;
        }
    }

    private AppInfo[] getInfos(String selection, String[] selectionArgs) {
        SQLiteDatabase db = helper.getWritableDatabase();
        String[] columns = {AppInfo.TID, AppInfo.APP_ID, AppInfo.VERSION, AppInfo.NAME, AppInfo.SHORT_NAME,
                AppInfo.DESCRIPTION, AppInfo.START_URL, AppInfo.TYPE,
                AppInfo.AUTHOR_NAME, AppInfo.AUTHOR_EMAIL, AppInfo.DEFAULT_LOCAL, AppInfo.BACKGROUND_COLOR,
                AppInfo.THEME_DISPLAY, AppInfo.THEME_COLOR, AppInfo.THEME_FONT_NAME, AppInfo.THEME_FONT_COLOR,
                AppInfo.INSTALL_TIME, AppInfo.BUILT_IN, AppInfo.REMOTE, AppInfo.LAUNCHER};
        Cursor cursor = db.query(ManagerDBHelper.APP_TABLE, columns,selection, selectionArgs,null,null,null);
        AppInfo infos[] = new AppInfo[cursor.getCount()];
        int count = 0;
        while (cursor.moveToNext()) {
            AppInfo info = new AppInfo();
            info.tid = cursor.getInt(cursor.getColumnIndex(AppInfo.TID));
            info.app_id = cursor.getString(cursor.getColumnIndex(AppInfo.APP_ID));
            info.version = cursor.getString(cursor.getColumnIndex(AppInfo.VERSION));
            info.name = cursor.getString(cursor.getColumnIndex(AppInfo.NAME));
            info.short_name = cursor.getString(cursor.getColumnIndex(AppInfo.SHORT_NAME));
            info.description = cursor.getString(cursor.getColumnIndex(AppInfo.DESCRIPTION));
            info.start_url = cursor.getString(cursor.getColumnIndex(AppInfo.START_URL));
            info.type = cursor.getString(cursor.getColumnIndex(AppInfo.TYPE));
            info.author_name = cursor.getString(cursor.getColumnIndex(AppInfo.AUTHOR_NAME));
            info.author_email = cursor.getString(cursor.getColumnIndex(AppInfo.AUTHOR_EMAIL));
            info.default_locale = cursor.getString(cursor.getColumnIndex(AppInfo.DEFAULT_LOCAL));
            info.background_color = cursor.getString(cursor.getColumnIndex(AppInfo.BACKGROUND_COLOR));
            info.theme_display = cursor.getString(cursor.getColumnIndex(AppInfo.THEME_DISPLAY));
            info.theme_font_name = cursor.getString(cursor.getColumnIndex(AppInfo.THEME_FONT_NAME));
            info.theme_font_color = cursor.getString(cursor.getColumnIndex(AppInfo.THEME_FONT_COLOR));
            info.theme_color = cursor.getString(cursor.getColumnIndex(AppInfo.THEME_COLOR));

            info.install_time = cursor.getLong(cursor.getColumnIndex(AppInfo.INSTALL_TIME));
            info.built_in = cursor.getInt(cursor.getColumnIndex(AppInfo.BUILT_IN));
            info.remote = cursor.getInt(cursor.getColumnIndex(AppInfo.REMOTE));
            info.launcher = cursor.getInt(cursor.getColumnIndex(AppInfo.LAUNCHER));
            infos[count++] = info;


            String[] args1 = {String.valueOf(info.tid)};

            String[] columns1 = {AppInfo.SRC, AppInfo.SIZES, AppInfo.TYPE};
            Cursor cursor1 = db.query(ManagerDBHelper.ICONS_TABLE, columns1,AppInfo.APP_TID + "=?", args1,null,null,null);
            while (cursor1.moveToNext()) {
                info.addIcon(cursor1.getString(cursor1.getColumnIndex(AppInfo.SRC)),
                        cursor1.getString(cursor1.getColumnIndex(AppInfo.SIZES)),
                        cursor1.getString(cursor1.getColumnIndex(AppInfo.TYPE)));
            }

            String[] columns2 = {AppInfo.PLUGIN, AppInfo.AUTHORITY};
            cursor1 = db.query(ManagerDBHelper.AUTH_PLUGIN_TABLE, columns2,AppInfo.APP_TID + "=?", args1,null,null,null);
            while (cursor1.moveToNext()) {
                info.addPlugin(cursor1.getString(cursor1.getColumnIndex(AppInfo.PLUGIN)), cursor1.getInt(cursor1.getColumnIndex(AppInfo.AUTHORITY)));
            }

            String[] columns3 = {AppInfo.URL, AppInfo.AUTHORITY};
            cursor1 = db.query(ManagerDBHelper.AUTH_URL_TABLE, columns3,AppInfo.APP_TID + "=?", args1,null,null,null);
            while (cursor1.moveToNext()) {
                info.addUrl(cursor1.getString(cursor1.getColumnIndex(AppInfo.URL)), cursor1.getInt(cursor1.getColumnIndex(AppInfo.AUTHORITY)));
            }

            String[] columns4 = {AppInfo.LANGUAGE, AppInfo.NAME, AppInfo.SHORT_NAME, AppInfo.DESCRIPTION, AppInfo.AUTHOR_NAME};
            cursor1 = db.query(ManagerDBHelper.LACALE_TABLE, columns4,AppInfo.APP_TID + "=?", args1,null,null,null);
            while (cursor1.moveToNext()) {
                info.addLocale(cursor1.getString(cursor1.getColumnIndex(AppInfo.LANGUAGE)),
                        cursor1.getString(cursor1.getColumnIndex(AppInfo.NAME)),
                        cursor1.getString(cursor1.getColumnIndex(AppInfo.SHORT_NAME)),
                        cursor1.getString(cursor1.getColumnIndex(AppInfo.DESCRIPTION)),
                        cursor1.getString(cursor1.getColumnIndex(AppInfo.AUTHOR_NAME)));
            }

            String[] columns5 = {AppInfo.NAME, AppInfo.VERSION};
            cursor1 = db.query(ManagerDBHelper.FRAMEWORK_TABLE, columns5,AppInfo.APP_TID + "=?", args1,null,null,null);
            while (cursor1.moveToNext()) {
                info.addFramework(cursor1.getString(cursor1.getColumnIndex(AppInfo.NAME)),
                        cursor1.getString(cursor1.getColumnIndex(AppInfo.VERSION)));
            }

            cursor1 = db.query(ManagerDBHelper.PLATFORM_TABLE, columns5,AppInfo.APP_TID + "=?", args1,null,null,null);
            while (cursor1.moveToNext()) {
                info.addFramework(cursor1.getString(cursor1.getColumnIndex(AppInfo.NAME)),
                        cursor1.getString(cursor1.getColumnIndex(AppInfo.VERSION)));
            }
        }
        return infos;
    }

    public AppInfo getAppInfo(String id) {
        String selection = AppInfo.APP_ID + "=? && " + AppInfo.LAUNCHER + "=0";
        String[] args = {String.valueOf(id)};
        AppInfo infos[] = getInfos(selection, args);
        if (infos.length > 0) {
            return infos[0];
        }
        else {
            return null;
        }
    }

    public AppInfo[] getAppInfos() {
        String selection = AppInfo.LAUNCHER + "=0";
        return getInfos(selection, null);
    }

    public AppInfo getLauncherInfo() {
        String selection = AppInfo.LAUNCHER + "=1";
        AppInfo infos[] = getInfos(selection, null);
        if (infos.length > 0) {
            return infos[0];
        }
        else {
            return null;
        }
    }


    public int updatePluginAuth(long tid, String plugin, int authority) {
        SQLiteDatabase db = helper.getWritableDatabase();
        ContentValues contentValues = new ContentValues();
        contentValues.put(AppInfo.AUTHORITY, authority);
        String where = AppInfo.APP_TID + "=? AND " + AppInfo.PLUGIN + "=?";
        String[] whereArgs= {String.valueOf(tid), plugin};
        int count = db.update(ManagerDBHelper.AUTH_PLUGIN_TABLE, contentValues, where, whereArgs );
        return count;
    }

    public int updateURLAuth(long tid, String url, int authority) {
        SQLiteDatabase db = helper.getWritableDatabase();
        ContentValues contentValues = new ContentValues();
        contentValues.put(AppInfo.AUTHORITY, authority);
        String where = AppInfo.APP_TID + "=? AND " + AppInfo.URL + "=?";
        String[] whereArgs= {String.valueOf(tid), url};
        int count =db.update(ManagerDBHelper.AUTH_URL_TABLE, contentValues, where, whereArgs );
        return count;
    }

    public int removeAppInfo(AppInfo info) {
        SQLiteDatabase db = helper.getWritableDatabase();
        String where = AppInfo.APP_TID + "=?";
        String[] whereArgs= {String.valueOf(info.tid)};
        int count = db.delete(ManagerDBHelper.AUTH_URL_TABLE, where, whereArgs);
        count = db.delete(ManagerDBHelper.AUTH_PLUGIN_TABLE, where, whereArgs);
        db.delete(ManagerDBHelper.ICONS_TABLE, where, whereArgs);
        db.delete(ManagerDBHelper.LACALE_TABLE, where, whereArgs);
        db.delete(ManagerDBHelper.FRAMEWORK_TABLE, where, whereArgs);
        db.delete(ManagerDBHelper.PLATFORM_TABLE, where, whereArgs);
        where = AppInfo.TID + "=?";
        count = db.delete(ManagerDBHelper.APP_TABLE, where, whereArgs);
        return count;
    }
}
